pkg/storegateway/batch_series.go

```diff
@@ -0,0 +1,255 @@
+	if s.i >= len(s.preloaded) {
+		return s.preload()
+	}
+	return true
+}
+
+func (s *batchedSeriesSet) preload() bool {
+	s.resetPreloaded()
+	s.iOffset += s.batchSize
+	if s.iOffset > len(s.postings) {
+		return false
+	}
+
+	end := s.iOffset + s.batchSize
+	if end > len(s.postings) {
+		end = len(s.postings)
+	}
+	nextBatch := s.postings[s.iOffset:end]
+
+	if err := s.indexr.PreloadSeries(s.ctx, nextBatch); err != nil {
+		s.err = errors.Wrap(err, "preload series")
+		return false
+	}
+
+	var (
+		symbolizedLset []symbolizedLabel
+		chks           []chunks.Meta
+	)
+	for _, id := range nextBatch {
+		ok, err := s.indexr.LoadSeriesForTime(id, &symbolizedLset, &chks, s.skipChunks, s.minTime, s.maxTime)
+		if err != nil {
+			s.err = errors.Wrap(err, "read series")
+			return false
+		}
+		if !ok {
+			// No matching chunks for this time duration, skip series and extend nextBatch
+			continue
+		}
+
+		lset, err := s.indexr.LookupLabelsSymbols(symbolizedLset)
+		if err != nil {
+			s.err = errors.Wrap(err, "lookup labels symbols")
+			return false
+		}
+
+		// Skip the series if it doesn't belong to the shard.
+		if s.shard != nil {
+			hash, ok := s.seriesHashCache.Fetch(id)
+			s.stats.seriesHashCacheRequests++
+
+			if !ok {
+				hash = lset.Hash()
+				s.seriesHashCache.Store(id, hash)
+			} else {
+				s.stats.seriesHashCacheHits++
+			}
+
+			if hash%s.shard.ShardCount != s.shard.ShardIndex {
+				continue
+			}
+		}
+
+		// Check series limit after filtering out series not belonging to the requested shard (if any).
+		if err := s.seriesLimiter.Reserve(1); err != nil {
+			s.err = errors.Wrap(err, "exceeded series limit")
+			return false
+		}
+
+		entry := seriesEntry{lset: lset}
+
+		if !s.skipChunks {
+			// Schedule loading chunks.
+			entry.refs = make([]chunks.ChunkRef, 0, len(chks))
+			entry.chks = make([]storepb.AggrChunk, 0, len(chks))
+			for j, meta := range chks {
+				// seriesEntry s is appended to res, but not at every outer loop iteration,
+				// therefore len(res) is the index we need here, not outer loop iteration number.
+				if err := s.chunkr.addLoad(meta.Ref, len(s.preloaded), j); err != nil {
+					s.err = errors.Wrap(err, "add chunk load")
+					return false
+				}
+				entry.chks = append(entry.chks, storepb.AggrChunk{
+					MinTime: meta.MinTime,
+					MaxTime: meta.MaxTime,
+				})
+				entry.refs = append(entry.refs, meta.Ref)
+			}
+
+			// Ensure sample limit through chunksLimiter if we return chunks.
+			if err := s.chunksLimiter.Reserve(uint64(len(entry.chks))); err != nil {
+				s.err = errors.Wrap(err, "exceeded chunks limit")
+				return false
+			}
+		}
+
+		s.preloaded = append(s.preloaded, entry)
+	}
+
+	if len(s.preloaded) == 0 {
+		return s.preload() // we didn't find any suitable series in this batch, try with the next one
+	}
+
+	if s.skipChunks {
+		storeCachedSeries(s.ctx, s.indexr.block.indexCache, s.indexr.block.userID, s.indexr.block.meta.ULID, s.matchers, s.shard, s.preloaded, s.logger)
```

Comment:
I'm not sure this is correct. `storeCachedSeries()` caches **all** series for a given block and label matchers, but since we're batching, we're not storing all series here just some of them.

Reply:
i missed this. Thanks for catching it